src/Avalonia.Visuals/Rendering/RenderLoop.cs

```diff
@@ -84,21 +85,35 @@ public void Remove(IRenderLoopTask i)
             }
         }
 
-        private async void TimerTick(TimeSpan time)
+        private void TimerTick(TimeSpan time)
         {
-            if (Interlocked.CompareExchange(ref inTick, 1, 0) == 0)
+            if (Interlocked.CompareExchange(ref _inTick, 1, 0) == 0)
             {
                 try
                 {
-                    if (_items.Any(item => item.NeedsUpdate))
+                    if (_items.Any(item => item.NeedsUpdate) &&
+                        Interlocked.CompareExchange(ref _inUpdate, 1, 0) == 0)
                     {
-                        await _dispatcher.InvokeAsync(() =>
+                        System.Diagnostics.Debug.WriteLine("Posted update");
+                        _dispatcher.Post(() =>
                         {
                             foreach (var i in _items)
                             {
-                                i.Update(time);
+                                if (i.NeedsUpdate)
+                                {
+                                    try
+                                    {
+                                        i.Update(time);
+                                    }
+                                    catch (Exception ex)
+                                    {
+                                        Logger.Error(LogArea.Visual, this, "Exception in render update: {Error}", ex);
+                                    }
+                                }
                             }
-                        }, DispatcherPriority.Render).ConfigureAwait(false);
+
+                            Interlocked.Exchange(ref _inUpdate, 0);
```

Comment:
We should wrap this in a try-finally block so an exception is thrown during the iteration (ex. someone adding a render task and modifying the list), we correctly mark that we are no longer in the update call.

Reply:
It might be worth to implement a collection with modification-safe iterators. There are way too many cases when modification during iteration scenario would be valid and defensive copy of the list would be too expensive.

Reply:
Should we just use a `for` loop with an indexer for now?

Reply:
Changed to use a `for` loop.